### PR TITLE
[FEMDEM] Reimplementing some processes from python to c++

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm reimplementing the DEM generation, kinematic coupling and Force transfer to the FEM. This change from python to c++ has reduced the calculation time about a 60% :)

### CODE REVIEW

**Comment by loumalouomega** [comment]
I need to leave now, I will check it later

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+array_1d<double,3> GenerateDemProcess::GetNodeCoordinates(
+    const NodeType& rNode
+    )
+{
+    array_1d<double,3> coordinates;
+    coordinates[0] = rNode.X();
+    coordinates[1] = rNode.Y();
+    coordinates[2] = rNode.Z();
+    return coordinates;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]: rNode.Coordinates()

Comment (loumalouomega) [reply applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]: And put it on the original place

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void UpdateDemKinematicsProcess::UpdateKinematics(
+    const NodeIteratorType& rNode,
+    NodeType& rDEMNode
+    )
+{
+    const array_1d<double,3> coordinates = this->GetNodeCoordinates(rNode);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp]:
```suggestion
    const array_1d<double,3>& coordinates = rNode.Coordinates();
```

Like this

Comment (loumalouomega) [reply applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp]: Of course r_, but that would break code

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Blocking for now, I will continue later

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    const NodeType& rNode2
+    )
+{
+    const double X1 = rNode1.X();
+    const double X2 = rNode2.X();
+    const double Y1 = rNode1.Y();
+    const double Y2 = rNode2.Y();
+    const double Z1 = rNode1.Z();
+    const double Z2 = rNode2.Z();
+    return std::sqrt(std::pow(X1-X2, 2.0) + std::pow(Y1-Y2, 2.0) + std::pow(Z1-Z2, 2.0));
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]:
```suggestion
    return std::sqrt(inner_prod(rNode1.Coordinates(), Node2.Coordinates()));
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    const NodeType& rNode1, 
+    const NodeType& rNode2
+    )
+{
+    const double X1 = rNode1.X();
+    const double X2 = rNode2.X();
+    const double Y1 = rNode1.Y();
+    const double Y2 = rNode2.Y();
+    const double Z1 = rNode1.Z();
+    const double Z2 = rNode2.Z();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp:133]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+double GenerateDemProcess::CalculateDistanceBetweenNodes(
+    const NodeType& rNode1, 
+    const NodeType& rNode2
+    )
+{
+    const double X1 = rNode1.X();
+    const double X2 = rNode2.X();
+    const double Y1 = rNode1.Y();
+    const double Y2 = rNode2.Y();
+    const double Z1 = rNode1.Z();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp:132]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+double GenerateDemProcess::CalculateDistanceBetweenNodes(
+    const NodeType& rNode1, 
+    const NodeType& rNode2
+    )
+{
+    const double X1 = rNode1.X();
+    const double X2 = rNode2.X();
+    const double Y1 = rNode1.Y();
+    const double Y2 = rNode2.Y();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp:131]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+/***********************************************************************************/
+
+double GenerateDemProcess::CalculateDistanceBetweenNodes(
+    const NodeType& rNode1, 
+    const NodeType& rNode2
+    )
+{
+    const double X1 = rNode1.X();
+    const double X2 = rNode2.X();
+    const double Y1 = rNode1.Y();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp:130]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+/***********************************************************************************/
+/***********************************************************************************/
+
+double GenerateDemProcess::CalculateDistanceBetweenNodes(
+    const NodeType& rNode1, 
+    const NodeType& rNode2
+    )
+{
+    const double X1 = rNode1.X();
+    const double X2 = rNode2.X();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp:129]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double GenerateDemProcess::CalculateDistanceBetweenNodes(
+    const NodeType& rNode1, 
+    const NodeType& rNode2
+    )
+{
+    const double X1 = rNode1.X();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp:128]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp (hunk 1 above)
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]:
```suggestion
    return rNode.Coordinates();
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+/***********************************************************************************/
+
+array_1d<double,3> GenerateDemProcess::GetNodeCoordinates(
+    const NodeType& rNode
+    )
+{
+    array_1d<double,3> coordinates;
+    coordinates[0] = rNode.X();
+    coordinates[1] = rNode.Y();
+    coordinates[2] = rNode.Z();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+/***********************************************************************************/
+/***********************************************************************************/
+
+array_1d<double,3> GenerateDemProcess::GetNodeCoordinates(
+    const NodeType& rNode
+    )
+{
+    array_1d<double,3> coordinates;
+    coordinates[0] = rNode.X();
+    coordinates[1] = rNode.Y();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+array_1d<double,3> GenerateDemProcess::GetNodeCoordinates(
+    const NodeType& rNode
+    )
+{
+    array_1d<double,3> coordinates;
+    coordinates[0] = rNode.X();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+array_1d<double,3> GenerateDemProcess::GetNodeCoordinates(
+    const NodeType& rNode
+    )
+{
+    array_1d<double,3> coordinates;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double GenerateDemProcess::GetMinimumValue(
+    const Vector& rValues
+    )
+{ // this method assumes that the Vector is NOT full of 0.0's
+    double aux = 1.0e10;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp:144]:
```suggestion
    double aux = std::numeric_limits<double>::max();
```

File: applications/FemToDemApplication/custom_processes/transfer_nodal_forces_to_fem.cpp [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+    const auto it_cond_begin = sub_model_conditions.ConditionsBegin();
+    //#pragma omp parallel for
+    for (int i = 0; i < static_cast<int>(sub_model_conditions.Conditions().size()); i++) {
+        auto it_cond = it_cond_begin + i;
+        auto& r_geometry = it_cond->GetGeometry();
+        const int id_node = r_geometry[0].Id();
+
+        if (mrModelPart.GetNode(id_node).GetValue(IS_DEM)) {
+            const array_1d<double,3>& dem_forces = mrDEMModelPart.GetNode(id_node).GetSolutionStepValue(TOTAL_FORCES);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/transfer_nodal_forces_to_fem.cpp:43]:
```suggestion
            const array_1d<double,3>& r_dem_forces = mrDEMModelPart.GetNode(id_node).GetSolutionStepValue(TOTAL_FORCES);
```

File: applications/FemToDemApplication/custom_processes/transfer_nodal_forces_to_fem.cpp [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+            it_cond->SetValue(FORCE_LOAD, dem_forces);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/transfer_nodal_forces_to_fem.cpp:44]:
```suggestion
            it_cond->SetValue(FORCE_LOAD, r_dem_forces);
```

File: applications/FemToDemApplication/custom_processes/transfer_nodal_forces_to_fem.cpp [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+void TransferNodalForcesToFem::Execute() 
+{
+	auto& sub_model_conditions = mrModelPart.GetSubModelPart("ContactForcesDEMConditions");
+
+    const auto it_cond_begin = sub_model_conditions.ConditionsBegin();
+    //#pragma omp parallel for
+    for (int i = 0; i < static_cast<int>(sub_model_conditions.Conditions().size()); i++) {
+        auto it_cond = it_cond_begin + i;
+        auto& r_geometry = it_cond->GetGeometry();
+        const int id_node = r_geometry[0].Id();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/transfer_nodal_forces_to_fem.cpp:40]: Geometry is always only one node?

File: applications/FemToDemApplication/custom_processes/transfer_nodal_forces_to_fem.cpp [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+        if (mrModelPart.GetNode(id_node).GetValue(IS_DEM)) {
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/transfer_nodal_forces_to_fem.cpp:42]:
```suggestion
        if (r_geometry[0].GetValue(IS_DEM)) {
```
Much faster

File: applications/FemToDemApplication/custom_processes/transfer_nodal_forces_to_fem.cpp (hunk 17 above)
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/transfer_nodal_forces_to_fem.cpp:40]:
```suggestion
```

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+void UpdateDemKinematicsProcess::Execute() 
+{
+    const auto it_node_begin = mrModelPart.NodesBegin();
+    //#pragma omp parallel for
+    for (int i = 0; i < static_cast<int>(mrModelPart.Nodes().size()); i++) {
+        auto it_node = it_node_begin + i;
+        if (it_node->GetValue(IS_DEM)) {
+            const int node_id = it_node->Id();
+            auto& associated_dem = mrDEMModelPart.GetNode(node_id);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp:39]:
```suggestion
            auto& r_associated_dem = mrDEMModelPart.GetNode(node_id);
```

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+            this->UpdateKinematics(it_node, associated_dem);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp:40]:
```suggestion
            this->UpdateKinematics(it_node, r_associated_dem);
```

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp (hunk 19 above)
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp:39]: This is slow...

Comment (loumalouomega) [reply applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp:39]: Is the node shared?

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+
+array_1d<double,3> UpdateDemKinematicsProcess::GetNodeCoordinates(
+    const NodeIteratorType& rNode
+    )
+{
+    array_1d<double,3> coordinates;
+    coordinates[0] = rNode->X();
+    coordinates[1] = rNode->Y();
+    coordinates[2] = rNode->Z();
+    return coordinates;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp]: Idem...

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+    const array_1d<double,3> coordinates = this->GetNodeCoordinates(rNode);
+    const array_1d<double,3> displacement = rNode->GetSolutionStepValue(DISPLACEMENT);
+    const array_1d<double,3> velocity = rNode->GetSolutionStepValue(VELOCITY);
+
+    auto& displ_dem = rDEMNode.GetSolutionStepValue(DISPLACEMENT);
+    displ_dem = displacement;
+    auto& vel_dem = rDEMNode.GetSolutionStepValue(VELOCITY);
+    vel_dem = velocity;
+
+    double& r_x_dem =  rDEMNode.X();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp]: Idem

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+void UpdateDemKinematicsProcess::UpdateKinematics(
+    const NodeIteratorType& rNode,
+    NodeType& rDEMNode
+    )
+{
+    const array_1d<double,3> coordinates = this->GetNodeCoordinates(rNode);
+    const array_1d<double,3> displacement = rNode->GetSolutionStepValue(DISPLACEMENT);
+    const array_1d<double,3> velocity = rNode->GetSolutionStepValue(VELOCITY);
+
+    auto& displ_dem = rDEMNode.GetSolutionStepValue(DISPLACEMENT);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp]: Missing r_

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+/***********************************************************************************/
+/***********************************************************************************/
+
+void UpdateDemKinematicsProcess::UpdateKinematics(
+    const NodeIteratorType& rNode,
+    NodeType& rDEMNode
+    )
+{
+    const array_1d<double,3> coordinates = this->GetNodeCoordinates(rNode);
+    const array_1d<double,3> displacement = rNode->GetSolutionStepValue(DISPLACEMENT);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp:54]: Missing &

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Comments

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]: it is not exactly the same operation... I want the modulus of the distance between them

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp (hunk 4 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/generate_dem_process.cpp:133]: in 3D i need to know the z components.. :)

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp (hunk 19 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp:39]: It is not shared... There are some cases that could exist a DEM node but the original fem node is erased. Maybe I can rethink this issue in the future :S

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp (hunk 24 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp:54]: In the begining I added the r_ and there was some memory corruption when reassigning. So I'll leave it that way

**Comment by AlejandroCornejo** [comment]
I've performed the changes @loumalouomega :) thx

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp (hunk 3 above)
Comment (loumalouomega) [reply applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]: In any case, replace 2.0 for 2, it is faster. And it is indeed faster my solution :P

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp (hunk 24 above)
Comment (loumalouomega) [reply applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp:54]: Are you removing the nodes?

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Minor request

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 25]
```diff
@@ -135,21 +135,6 @@ double GenerateDemProcess::CalculateDistanceBetweenNodes(
     return std::sqrt(std::pow(X1-X2, 2.0) + std::pow(Y1-Y2, 2.0) + std::pow(Z1-Z2, 2.0));
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]:
```suggestion
    return std::sqrt(std::pow(X1-X2, 2) + std::pow(Y1-Y2, 2) + std::pow(Z1-Z2, 2));
```

**Comment by AlejandroCornejo** [comment]
Done @loumalouomega :)

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**